Tesla Counterplay.

jack-avery opened the pull request:
## About The Pull Request

- Grounding rods now pull in the Tesla. The nearest grounding rod is prioritized.
- If the Tesla has one or more miniballs, the grounding rod will be consumed and take some of the Tesla's energy with it (equivalent to two miniballs)
- If the Tesla has no miniballs (energy of 12 or below), it will continue to be anchored nearby a grounding rod (if there is one).
- Teslas will now fizzle once their energy reaches 0. There is an admin log for this.
- The Tesla 'contained' variable now functions as expected, with it's own unique check. Works guaranteed with containments up to 9x9 tiles of space (guaranteed on roundstart setups for Delta and Lambda stations).

## Why It's Good For The Game

tesla counterplay

## Changelog
:cl:
tweak: tesla movement priorities
add: tesla counterplay
add: tesla containment check (containment variable now usable)
/:cl:

<details>
  <summary>old description</summary>
  Grounding rods now attract the Tesla. Exactly how much is luck-based, but you can expect it to gravitate generally towards a rod.
Grounding rods, when zapped by a Tesla with miniballs, will eat a chunk of energy from the Tesla equivalent to two miniballs.
This gives a Tesloose some considerable counterplay, and since this is unmapped territory, we'll have to see whether this needs buffing or not.
Testmerge suggested to see if this is powerful enough before a fullmerge, though the situations under which this occurs are generally rare.

Edit, for more details on the effects:
A Tesla, when surrounded by containment fields on all sides within a radius 10 tiles, is considered contained. If your Tesla containment is 10x10 (or smaller) this is fine (and this is how I most often see the Tesla contained). But, if the Tesla containment is larger than 10x10, the Tesla has a decent chance to move out of range and be considered uncontained during this tick, destroying one (or more) of the nearby grounding rods, opening up the rest of the area to be fried.
</details>

jack-avery committed:
init

first attempt

jack-avery committed:
it likes moving grounding rod

now it actually gets the dir instead of just adding the rod itself to the pick

also moved declaration of move_dir because dm was freaking out because i had two declarations

jack-avery committed:
tesla counterplay??

just put out some grounding rods and it'll eat up the rods at expense of some energy

jack-avery committed:
oops! no grounding rods

now it (hopefully) won't delete rods from within containment

Putnam3145 commented on code/modules/power/tesla/energy_ball.dm:
```diff
@@ -31,6 +31,7 @@
 	var/produced_power
 	var/energy_to_raise = 32
 	var/energy_to_lower = -20
+	var/rodtarget
```
```suggestion
	var/obj/machinery/power/grounding_rod/rodtarget
```

Putnam3145 commented on code/modules/power/tesla/energy_ball.dm:
```diff
@@ -91,8 +92,21 @@
 	var/move_bias = pick(GLOB.alldirs)
+	var/move_dir
+	var/rods = GLOB.grounding_rods
+	if(rods) // grounding rods pull the tesla ball, picks the nearest one
```
this is unnecessary, if GLOB doesn't have grounding_rods it'll yell anyway afaik

jack-avery replied:
i did this for readability more than anything, would removing the if(rods) and replacing the for(in rods) with (in GLOB) work?

Putnam3145 commented on code/modules/power/tesla/energy_ball.dm:
```diff
@@ -91,8 +92,21 @@
+	var/rods = GLOB.grounding_rods
+	if(rods) // grounding rods pull the tesla ball, picks the nearest one
+		for(var/rod in rods) 
+			if(!rodtarget)
```
```suggestion
			if(!rodtarget || get_dist(src,rod)<get_dist(src,rodtarget))
```

Putnam3145 commented on code/modules/power/tesla/energy_ball.dm:
```diff
@@ -91,8 +92,21 @@
+				rodtarget=rod
+			
+			if(get_dist(src,rod)<get_dist(src,rodtarget))
+				rodtarget=rod
```
if committing above suggestion, remove these two lines

Putnam3145 commented on code/modules/power/tesla/energy_ball.dm:
```diff
@@ -91,8 +92,21 @@
+
 	for(var/i in 0 to move_amount)
-		var/move_dir = pick(GLOB.alldirs + move_bias) //ensures large-ball teslas don't just sit around
+		if(rods)
```
```suggestion
		if(rodtarget)
```

jack-avery committed:
Update code/modules/power/tesla/energy_ball.dm

Co-authored-by: Putnam3145 <putnam3145@gmail.com>

jack-avery committed:
Update code/modules/power/tesla/energy_ball.dm

Co-authored-by: Putnam3145 <putnam3145@gmail.com>

jack-avery committed:
Update code/modules/power/tesla/energy_ball.dm

Co-authored-by: Putnam3145 <putnam3145@gmail.com>

jack-avery committed:
if committing above suggestion, remove these two lines

ok done

jack-avery committed:
if(rods) literally just eats disk space and nothing more

no longer. 300 bytes saved

jack-avery committed:
tesla can now be contained

for whatever reason, the contained variable was unused for teslas, but now it is
unintentional side-effect may be that grounding rods for tesla containments with a larger than 1x1 tile space might get eaten (untested but assumed)

jack-avery committed:
create unique containment check for the tesla

it now functions as normal. it only checks for containment in a 10 tile radius (this should work fine for the default setups of Delta and Lambda).

jack-avery committed:
remove excess whitespace

title

jack-avery committed:
shuffle containment checking for optimization

now it should be checking less (tested, functionally identical to before)

jack-avery committed:
why is this here

no really how the fuck did this happen

jack-avery committed:
teslas now fizzle at 0 energy and adds admin log for fizzles

fixing the tesla engine 100% speedrun world record

jack-avery committed:
remove redundant separate tesla log and merge it into the singulo log

also formatting because i hate inconsistent whitespace
